feat: add Azure Active Directory authentication guide (#4964)
cypress-io/cypress-documentation: docs/guides/end-to-end-testing/testing-strategies/azure-active-directory-authentication.mdx

    cy.visit('http://localhost:3000')
  })

  it('verifies the user logged in has the correct name', () => {
    cy.get('#table-body-div td:contains("name") + td').should(
      'contain',
      `${Cypress.env('aad_name')}`
    )
  })

  it('verifies the user logged in has the correct preferred name', () => {
    cy.get('#table-body-div td:contains("preferred_username") + td').should(
      'contain',
      `${Cypress.env('aad_username')}`
    )
  })
})
```

<DocsVideo src="https://vimeo.com/789095126"></DocsVideo>

We now have working authentication and tests! But we are logging in before every
test, which is not only time consuming, but also can lead to API rate limiting
due to the number of requests.

For this, we can refactor our login command to take advantage of
[`cy.session()`](/api/commands/session) to store our logged in user's tokens and/or cookies
so we don't have to reauthenticate before every test.

```js
Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
  cy.session(
    `aad-${username}`,
    () => {
      const log = Cypress.log({
        displayName: 'Azure Active Directory Login',
        message: [`🔐 Authenticating | ${username}`],

>         // @ts-ignore

Reviewer: ```suggestion
```

Author: addressed in 7e426fc8a

---
title: Azure Active Directory Authentication
slug: /guides/end-to-end-testing/azure-active-directory-authentication
---

:::info

## <Icon name="graduation-cap"></Icon> What you'll learn

- Log in to
  [Azure Active Directory](https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/active-directory-whatis)
  through the UI with [`cy.origin()`](/api/commands/origin)

:::

:::caution

This guide is designed for testing against a Single Page Application (SPA) that
uses

> [Azure Active Directory](https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/active-directory-whatis)

Reviewer: ```suggestion
[Azure Active Directory](https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/active-directory-whatis) (AAD)
```
you use this abreviate below without calling out what it stands for

Author: I thought I had it in the Microsoft AAD App set up but I reference AAD first before actually defining it 🤦🏻

Author: addressed in 7e426fc8a

to authenticate users. For this guide, the Microsoft Authentication Library
[`@azure/msal-browser`](https://www.npmjs.com/package/@azure/msal-browser)

> package is used by the webapp to broker this authentication.

Reviewer: ```suggestion
package is used by the web app to broker this authentication.
```

Author: addressed in 7e426fc8a

> This guide can also serve as a foundation for testing other webapps with Cypress

Reviewer: ```suggestion
This guide can also serve as a foundation for testing other web apps with Cypress
```

foundation or starting point?

Author: addressed in 7e426fc8a

that use Azure Active Directory services with other frameworks, such as React,
Angular, or Vue.

:::

## Microsoft AAD Application Setup

For this guide, we are mainly going to focus on setting up Cypress to test

> against an Azure Active Directory (AAD) Webapp. Please clone the

Reviewer: ```suggestion
against an Azure Active Directory Webapp. Please clone the
```

Author: addressed in 7e426fc8a

[Microsoft Identity Javascript Tutorial](https://github.com/Azure-Samples/ms-identity-javascript-tutorial/tree/c1956b658efa331bb5df11a0038ad32d12dad3ce/1-Authentication/1-sign-in)
and follow the steps to set up your application.

> Once set up, you will need to do a few things to the `index.html` file:

Reviewer: ```suggestion
Once set up, you will need to modify a few things in the `App/index.html` file:
```
looking the code you linked, index.html is not at the root level.

Author: addressed in 7e426fc8a

> - Remove any integrity attributes inside `<script>` tags. This is to prevent any

Reviewer: ```suggestion
- Remove any [`integrity`](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) attributes inside `<script>` tags. This is to prevent any
```

Author: addressed in 7e426fc8a

  SRI attribute mismatches, which can happen since Cypress must rewrite any
  [framebusting code](/guides/references/configuration#modifyObstructiveCode).

>   This is recommended only when an application is under test.

Reviewer: If we remove these while testing, does this impact the integrity of the e2e tests? is it possible to leave these integrity attributes?

Author: When you say "impact the integrity of the e2e test" are you referring that it should throw an error if the integrity is there or that users don't have a solid test because they removed the integrity? Or none of those thing? 😅 

Right now I don't think there is a way to keep the integrity values in the app `index.html` since we don't strip integrity for first party attributes:

Heres the code in the proxy  in `response-middleware` that enforces this
![](https://user-images.githubusercontent.com/3980464/212986925-b729d357-a120-4c19-b8dc-e5aa90e5087e.png)

I feel like this is something users would be unwilling to do, and a better option we could provide is have a unique flag for SRI stripping and when its turned on it strips SRI for everything. Then the user wouldn't have to modify their index file here and Cypress would just automatically do it on opt in.

Reviewer: yeah the first statement

- Uncomment the `authRedirect.js` `<script>` tag and comment out the
  `authPopup.js` `<script>` tag. Authentication
  [Pop ups](/guides/references/trade-offs#Permanent-trade-offs) will not work
  inside of Cypress.

Inside `server.js`, you will see a reference to `express-rate-limit`, which
limits each window to 100 requests per 15 minutes. Under test, your application
will likely exceed this as you will be reloading your application and making
requests to broker/verify authentication. For this demo, you can either

- remove the rate limiting code (recommended for demo purposes only).
- Implement a strategy to increase the rate limit under test.

When finished, your SPA should be running on `http://localhost:3000` and be

> properly configured to run with Azure Active Directory.

Reviewer: ```suggestion
properly configured to run with Azure Active Directory and Cypress.
```

Author: addressed in 7e426fc8a

## Configuring Cypress to use Microsoft AAD

To have access to test user credentials within our tests, we need to configure
Cypress to use the
[Azure Active Directory](https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/active-directory-whatis)
user credentials for a given user in your tenant. These user credentials should
be set in the `cypress.env.json` file inside the root directory of your project.

```json
// cypress.env.json
{
  "aad_username": "AAD_USERNAME",
  "aad_password": "AAD_PASSWORD",
  "aad_name": "AAD_NAME"
}
```

> Also, to authenticate with Microsoft or Azure Active Directory, you will need to

Reviewer: ```suggestion
Also, to authenticate with Azure Active Directory, you will need to
```

Author: addressed in 7e426fc8a

enable the
[`experimentalModifyObstructiveThirdPartyCode`](/guides/guides/web-security#Modifying-Obstructive-Third-Party-Code)

> configuration option. If not enabled, your authentication workflow will enter an

Reviewer: ```suggestion
configuration option in the e2e configuration. If not enabled, your authentication workflow will enter an
```

Author: addressed in 7e426fc8a

This command will use [`cy.origin()`](/api/commands/origin) to

1. Navigate to the Azure Active Directory login page on
   `login.microsoftonline.com`
2. Input user credentials
3. Sign in and redirect back to the demo application
4. Cache the results with [`cy.session()`](/api/commands/session)

```js
// cypress/support/e2e.ts

function loginViaAAD(username: string, password: string) {
  cy.visit('http://localhost:3000/')
  cy.get('button#signIn').click()

  // Login to your AAD tenant.
  cy.origin(
    'login.microsoftonline.com',
    {
      args: {
        username,
      },
    },
    ({ username }) => {
      cy.get('input[type="email"]').type(username, {
        log: false,
      })
      cy.get('input[type="submit"]').click()
    }
  )

  // depending on the user and how they are registered with Microsoft, the origin may go to live.com
  cy.origin(
    'login.live.com',
    {
      args: {
        password,
      },
    },
    ({ password }) => {
      cy.get('input[type="password"]').type(password, {
        log: false,
      })
      cy.get('input[type="submit"]').click()

>       cy.get('#idBtn_Back').click()

Reviewer: whats this button? 👀

Author: It's the no button for "Stay signed in?". Doesn't have too much of an effect here its mainly to complete the login flow
![](https://user-images.githubusercontent.com/3980464/212986116-903c38d9-b267-4951-9893-5c8e3b76abb6.png)

    }
  )

  // Ensure Microsoft has redirected us back to the sample app with our logged in user.
  cy.url().should('equal', 'http://localhost:3000/')
  cy.get('#welcome-div').should(
    'contain',
    `Welcome ${Cypress.env('aad_username')}!`
  )
}

Cypress.Commands.add('loginToAAD', (username: string, password: string) => {

>   const log = Cypress.log({

Reviewer: We really should not be sharing these examples in Guides IMO because it requires the user to clean these up themselves if a failure occurs. Typically these will be left as pending when a failure occurs.

Author: Do you mean the added log message? I am not a big fan of it either for that, but mostly because it just adds code that really isn't necessary for understanding how this needs to work. The only reason I keep it here is to be consistent with the other guides. I wonder if we should do a pass at removing that in the docs, but keep it in the RWA if applicable?

Author: I'll likely keep it in, but I don't like the pattern in our guides. I think in the future we should remove it from the others as well, even programmatic login

    displayName: 'Azure Active Directory Login',
    message: [`🔐 Authenticating | ${username}`],

>     // @ts-ignore

Reviewer: ```suggestion
```

Author: addressed in 7e426fc8a

  log.end()
})
```

Now, we can use our `loginToAAD` command in the test. Below is our test to login
as a user via Azure Active Directory and run a basic sanity check.

```js
describe('Azure Active Directory Authentication', () => {
  beforeEach(() => {
    // log into Azure Active Directory through our sample SPA using our custom command
    cy.loginToAAD(Cypress.env('aad_username'), Cypress.env('aad_password'))
    cy.visit('http://localhost:3000')
  })

  it('verifies the user logged in has the correct name', () => {
    cy.get('#table-body-div td:contains("name") + td').should(
      'contain',
      `${Cypress.env('aad_name')}`
    )
  })

  it('verifies the user logged in has the correct preferred name', () => {
    cy.get('#table-body-div td:contains("preferred_username") + td').should(
      'contain',
      `${Cypress.env('aad_username')}`
    )
  })
})
```

<DocsVideo src="https://vimeo.com/789095126"></DocsVideo>

We now have working authentication and tests! But we are logging in before every
test, which is not only time consuming, but also can lead to API rate limiting
due to the number of requests.

For this, we can refactor our login command to take advantage of

> [`cy.session()`](/api/commands/session) to store our logged in user so we don't

Author: addressed in 7e426fc8a

have to reauthenticate before every test.

```js
Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
  cy.session(
    `aad-${username}`,
    () => {
      const log = Cypress.log({
        displayName: 'Azure Active Directory Login',
        message: [`🔐 Authenticating | ${username}`],
        // @ts-ignore
        autoEnd: false,
      })

      log.snapshot('before')

      loginViaAAD(username, password)

      log.snapshot('after')
      log.end()
    },
    {
      validate: () => {
        // this is a very basic form of session validation for this demo.
        // depending on your needs, something more verbose might be needed
        cy.visit('http://localhost:3000')
        cy.get('#welcome-div').should(
          'contain',
          `Welcome ${Cypress.env('aad_username')}!`
        )

>       },

Author: Not a lot of great options for Oauth it looks like. I looked at some of the values in cookies, local, and session storage, but couldnt find anything deterministic for what would guarantee the user is logged it. An approach I thought about using here was using the AAD library to figure out who is logged into the session and verifying the user that way, but I figured that might be too much for the simple guide. What do you think? Should we opt for that approach to avoid the visit?

>       },

Reviewer: It's hard to say - this type of validation is "heavier" and "slower" for users but might be "safest" for the longevity of the guide

})
```

<DocsVideo src="https://vimeo.com/789095126"></DocsVideo>

We now have working authentication and tests! But we are logging in before every
test, which is not only time consuming, but also can lead to API rate limiting
due to the number of requests.

For this, we can refactor our login command to take advantage of
[`cy.session()`](/api/commands/session) to store our logged in user's tokens
and/or cookies so we don't have to reauthenticate before every test.

```js
Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
  cy.session(
    `aad-${username}`,
    () => {
      const log = Cypress.log({
        displayName: 'Azure Active Directory Login',
        message: [`🔐 Authenticating | ${username}`],
        // @ts-ignore
        autoEnd: false,
      })

      log.snapshot('before')

      loginViaAAD(username, password)

      log.snapshot('after')
      log.end()
    },
    {
      validate: () => {
        // this is a very basic form of session validation for this demo.
        // depending on your needs, something more verbose might be needed
        cy.visit('http://localhost:3000')
        cy.get('#welcome-div').should(
          'contain',

>           `Welcome ${Cypress.env('aad_username')}!`

Reviewer 2: this feels like a funky use for the validate function. Whats it suppose to be doing here? If the idea is to get it to wait for the values to be ready before the session is serialized I would move the `cy.get('#welcome-div')` command to the bottom of the cy.session setup method.

Author: mainly the goal is to make sure the session is rehydrated correctly with the correct user logged in when the session is reinvoked. Moving to the end of the setup method method would mean it would only work the first time
